Unified safe list

iamacook opened the pull request:
## What it solves
Resolves #2617

## How this PR fixes it

## How to test it

## Screenshots

katspaugh commented on src/components/AppLayout/Header/components/NetworkSelector.tsx:
```diff
@@ -1,5 +1,4 @@
-import React, { ReactElement, useRef, Fragment, useCallback } from 'react'
-import { useDispatch } from 'react-redux'
+import React, { ReactElement, useRef, Fragment } from 'react'
```
Please pull from dev, React now shouldn't be imported.

iamacook replied:
I have added the rule to eslint now and removed all instances of `react` imports

katspaugh commented on src/components/SafeListSidebar/SafeList/SafeListItem.tsx:
```diff
@@ -0,0 +1,76 @@
+}: Props): React.ReactElement => {
+  const history = useHistory()
+  const safeName = useSelector((state) => addressBookEntryName(state, { address }))
+  const isSameAddress = sameAddress(currentSafeAddress, address)
```
`isCurrentSafe`?

katspaugh commented on src/components/SafeListSidebar/SafeList/SafeListItem.tsx:
```diff
@@ -0,0 +1,76 @@
+
+  return (
+    <ListItem button onClick={handleOpenSafe} ref={safeRef}>
+      <StyledIcon type="check" size="md" color="primary" isSameAddress={isSameAddress} />
```
In the context of this icon, the prop should be probably called `checked`, not `isSameAddress`.

katspaugh commented on src/components/SafeListSidebar/SafeList/SafeListItem.tsx:
```diff
@@ -0,0 +1,76 @@
+      <EthHashInfo hash={address} name={safeName} showAvatar shortenHash={4} />
+      <ListItemSecondaryAction>
+        {ethBalance ? (
+          `${formatAmount(ethBalance)} ETH`
```
`ETH` is only on mainnet/rinkeby.
You need to grab the native coin name from the network config.

dasanra commented on src/components/SafeListSidebar/SafeList/SafeListItem.tsx:
```diff
@@ -0,0 +1,76 @@
+import styled from 'styled-components'
+import { formatAmount } from 'src/logic/tokens/utils/formatAmount'
+import { useSelector } from 'react-redux'
+import { addressBookEntryName } from 'src/logic/addressBook/store/selectors'
```
Another convention we follow is to put external dependencies on top of the file and separated with one blank line from internal imports, components...

Could you fix this and apply the same from now? Don't worry if you have more files where you already did this, we will refactor them when we find them :wink:

iamacook replied:
I've updated all the files that have changed in this PR to follow this.

katspaugh commented on src/components/SafeListSidebar/SafeList/index.tsx:
```diff
@@ -45,70 +42,96 @@ const useStyles = makeStyles({
+  const currentSafeAddress = useSelector(safeAddressFromUrl)
+  const safes = useSelector(sortedSafeListSelector).filter((safe) => !safe.loadedViaUrl)
+  const ownedSafes = useOwnerSafes()
+  const safesNotAdded = ownedSafes.filter((address) => !isAddressAdded(safes, address))
```
It seems logical to only display the not added safes but we specifically made it display _all_ the owned safes as per CEO's request.

iamacook replied:
Do we want to show anything underneath the network heading, or just leave it blank? Maybe add Safe perhaps?

katspaugh replied:
The owned safes menu should just remain as it is.

iamacook replied:
![image](https://user-images.githubusercontent.com/20442784/133757581-e6e65645-0f11-4a00-9cd1-d53c14a458a2.png)

Like this?

iamacook replied:
Or perhaps with a message along the lines of: `No added Safes on ${network}`

katspaugh replied:
My point was that it should display _all_ owned safes, not just not added. See #2657 for context.

iamacook replied:
All safes are shown now.

katspaugh commented on src/logic/addressBook/utils/v2-migration.ts:
```diff
@@ -12,6 +12,8 @@ interface StorageConfig {
   namespaceSeparator: string
 }
 
+export const getSafesKey = (network: string) => `_immortal|v2_${network}__SAFES`
```
This migration script will be deleted soon, please don't rely on it in the safe list.

iamacook replied:
Shall I move this function to `SafeListItem` then?

katspaugh commented on src/components/SafeListSidebar/SafeList/index.tsx:
```diff
@@ -45,70 +42,96 @@ const useStyles = makeStyles({
+  let otherSafes: SafeRecordWithNames[] = []
+
+  const name = getNetworkName(networkId)
+  const safeString = localStorage.getItem(getSafesKey(name))
```
This should be using `loadFromStorage`. Please don't hack the localStorage directly.

iamacook replied:
I have had to modify `loadFromStorage` to allow me to use it:

![image](https://user-images.githubusercontent.com/20442784/133748349-e67e2b7a-cb47-442a-bb51-1caa0ef2879e.png)

![image](https://user-images.githubusercontent.com/20442784/133748404-8c1b6ac8-7b89-4f34-9d36-ebb603ea2b30.png)

What are your thoughts?

katspaugh replied:
Sounds good, we'll probably need `loadFromStorage` to have this ability in other places.

katspaugh commented on src/logic/config/utils/index.ts:
```diff
@@ -0,0 +1,11 @@
+  setNetworkId(getNetworkName(id))
+  const safeConfig = makeNetworkConfig(getConfig())
+  store.dispatch(configStore(safeConfig))
+}
```
@dasanra I think you were planning to do it via Redux?

iamacook replied:
I abstracted this from four positions. If we want to add it to Redux, it would be nice if @dasanra could lend a hand.

dasanra commented on package.json:
```diff
@@ -39,9 +39,9 @@
-    "start-bsc": "REACT_APP_LATEST_SAFE_VERSION='1.3.0' REACT_APP_NETWORK='bsc' yarn start",
+    "start-mainnet": "yarn start REACT_APP_NETWORK=mainnet",
+    "start-polygon": "yarn start REACT_APP_LATEST_SAFE_VERSION='1.3.0' REACT_APP_NETWORK=polygon",
+    "start-bsc": "yarn start REACT_APP_LATEST_SAFE_VERSION='1.3.0' REACT_APP_NETWORK='bsc'",
```
What is the motivation for this change?

Linux environments don't accept ENV parameters after the command so this won't work

mikheevm replied:
I can feel Bill Gates influence in this change

iamacook replied:
Windows machines need them after. Shall I create a `*-win` command for all the Windows users?

katspaugh replied:
This won't work on Mac either.

Maybe this'll work universally?

```
export REACT_APP_NETWORK=mainnet; yarn start
```

katspaugh replied:
Also, I assume windows users can run a Linux compatibility layer like WSL/Cygwin/MinGW.

iamacook replied:
I'll revert these changes and look into the best option for Windows as it's just me with a Windows machine.

iamacook replied:
> 
> 
> This won't work on Mac either.
> 
> Maybe this'll work universally?
> 
> ```
> export REACT_APP_NETWORK=mainnet; yarn start
> ```

As this didn't work and I'll look into WSL etc.

dasanra replied:
We should use `cross-env` then in this command

dasanra replied:
Try setting them like this: 

```
"start-mainnet": "cross-env REACT_APP_NETWORK=mainnet yarn start",
"start-polygon": "cross-env REACT_APP_LATEST_SAFE_VERSION='1.3.0' REACT_APP_NETWORK=polygon yarn start",
"start-bsc": "cross-env REACT_APP_LATEST_SAFE_VERSION='1.3.0' REACT_APP_NETWORK='bsc' yarn start",
 ```

mikheevm replied:
I think using WSL is a better option (for us) because I know some commands in other gnosis' packages heavily depend on the shell (scripts, etc.)

Richard knows how to set up it properly, and you can ask him

iamacook replied:
I've added `cross-env` for the time being.

iamacook commented:
Quick question before this (hopefully) makes it through review. Do we want to display 'UNKNOWN' as the default name for unnamed Safes in the unified list? Or only show the address? I fixed the selector in my hook abstraction and it now shows only the name when one exists. @katspaugh

katspaugh commented:
I can't test this branch, the wallet isn't connecting. A merge issue?

katspaugh commented:
Clicked around a bit.
We should probably display the All Owned Safes menu for the current network, if there are any owned safes.

iamacook commented:
> 
> 
> Clicked around a bit.
> We should probably display the All Owned Safes menu for the current network, if there are any owned safes. Even if no safes were _added_ on this network.

What do you think about this? @katspaugh

![image](https://user-images.githubusercontent.com/20442784/134471065-2341bee8-f955-4720-8867-5de5d7e0bc21.png)

For context:
- I am connected to EWC, where I own/have added no Safes.
- On Rinkeby I own 4, but the 3 I have added are shown because I can access them via `localStorage`.
- No other networks are shown because I have added no Safes to any others.

iamacook commented:
> 
> 
> Clicked around a bit.
> We should probably display the All Owned Safes menu for the current network, if there are any owned safes. Even if no safes were _added_ on this network.

![image](https://user-images.githubusercontent.com/20442784/134473351-afd24674-8f14-4189-a1bf-0fba264fa918.png)

And when Safes exist on the current network.

katspaugh commented on src/components/SafeListSidebar/SafeList/index.tsx:
```diff
@@ -46,70 +42,70 @@ const useStyles = makeStyles({
+                  {...safe}
+                />
+              ))}
+              {(ownedSafes.length > 0 || isCurrentNetwork) && (
```
We don't want to show this menu if there are no owned safes.

To summarize:

* If there are added safes, show them
* If there are owned safes on the current network, show them
* If there are neither added nor owned safes, hide the network section completely

iamacook replied:
![image](https://user-images.githubusercontent.com/20442784/134546988-c8e7bdf7-12b2-4b80-9610-10b425ced705.png)

Here's a screenshot of my most recent commit

katspaugh reviewed (approved):
👍

katspaugh commented on package.json:
```diff
@@ -39,9 +39,9 @@
-    "start-bsc": "REACT_APP_LATEST_SAFE_VERSION='1.3.0' REACT_APP_NETWORK='bsc' yarn start",
+    "start-mainnet": "cross-env REACT_APP_NETWORK=mainnet yarn start",
+    "start-polygon": "cross-env REACT_APP_LATEST_SAFE_VERSION='1.3.0' REACT_APP_NETWORK=polygon yarn start",
+    "start-bsc": "cross-env REACT_APP_LATEST_SAFE_VERSION='1.3.0' REACT_APP_NETWORK='bsc' yarn start",
```
`cross-env` is probably an npm package und thus needs to be installed?

iamacook replied:
I installed WSL a couple days ago, so I will revert these changes because they only applied to me

iamacook replied:
Coming back to this briefly, it is already installed for a few other commands.

francovenica commented:
- [ ] I cannot see my owned safes in a clean env. I'm connected with a wallet that has 600+ safes on rinkeby
As I understand by the comments, I should see the "Rinkeby" tag and bellow it the "Safes owned on Rinkeby" menu that display those safes
![image](https://user-images.githubusercontent.com/4503659/134679794-7ad4c0cf-c990-46d7-a984-5e69c78bec43.png)

- [ ] Once I added a localStorage key/value for rinkeby the list of "Safes owned on Rinkeby" showed up just fine, but is displayed by default. It should be collapsed by default

- [ ] If I change networks with the selector on the top right corner the safes disappear from the list. I don't know if for this PR is intended for that selector to work with the safe list, let me know if that is outside of the scope of the ticket
![09-24-2021_x(3048)](https://user-images.githubusercontent.com/4503659/134682071-c1e513dd-c1b3-4bdf-9256-1b581173cf2c.gif)

- [ ] The sidebar scrolls to the current safe, but instead of going to the network section where the safe is when is stored, is going to the "Safes Owned in NETWORK" list.
In this gif you can see how I'm taken to the "Safes owned" list instead of the main network safes list I have
![09-24-2021_x(3050)](https://user-images.githubusercontent.com/4503659/134700949-eb52a40b-146a-47ca-85fe-56160532010a.gif)

iamacook commented:
> If I change networks with the selector on the top right corner the safes disappear from the list. I don't know if for this PR is intended for that selector to work with the safe list, let me know if that is outside of the scope of the ticket

With regards to your third point, I cannot replicate it. Although, I think by fixing the first two points, it may have fixed it:
![unified-safes-remain](https://user-images.githubusercontent.com/20442784/134881805-530a54ad-5ded-405e-8018-9a78831dee62.gif)

francovenica commented:
The issues with the "Safes owned" menu not showing when I don't have safes of that network in the storage was solved
Same with the issue of that menu being displayed by default instead of being collapsed.

Issues:
- [ ] The scroll issue is still there. It keeps scrolling to the "Safes owned". Is easier to reproduce with a lot of safes:
![09-27-2021_x(3066)](https://user-images.githubusercontent.com/4503659/134914149-cd29cd9d-0e19-4161-bf11-df269ddcde7f.gif)

This file has a ton of rinkeby safes, it will be easier to reproduce with it:
[prodRinkeby_09-14-2021.zip](https://github.com/gnosis/safe-react/files/7236429/prodRinkeby_09-14-2021.zip)

- [ ] The Issue 3 seems to happen only when you are disconnected from the app: 
![09-27-2021_x(3064)](https://user-images.githubusercontent.com/4503659/134913444-f4f22be7-bf20-4f2c-852f-69f04db7832a.gif)

katspaugh commented:
The margin between network sections seems a bit too large. Probably the bottom margin of the Owned Safes container?

katspaugh commented:
The Owned Safes seem to have lost a piece of functionality from the original one.
When you open a safe from that list, the menu should stay expanded and scroll to that item.

katspaugh reviewed (approved):
Works perfectly now 👍

francovenica commented:
The 2 issues left were fixed fine
Tested the comment Ivan made:
If the you are in a safe and that safe was never added the "Safe owned" menu opens on its own and you scroll automatically to it. If you add that safe to the safe sidebar now it will scroll to "safe in network" menu instead of the "Safe owned" menu, which is correct

It looks good to me
